src/Shared/PackageManagers/BaseProjectManager.cs

```diff
@@ -310,6 +310,19 @@ public virtual Task<IEnumerable<string>> EnumerateVersionsAsync(PackageURL purl,
         {
             throw new NotImplementedException("BaseProjectManager does not implement EnumerateVersions.");
         }
+        
+        /// <summary>
+        /// Enumerates all possible versions of the package identified by purl, in descending order with the time they were published at.
+        /// </summary>
+        /// <remarks>The latest version is always first, then it is sorted by SemVer in descending order.</remarks>
+        /// <param name="purl">Package URL specifying the package. Version is ignored.</param>
+        /// <param name="useCache">If the cache should be used when looking for the versions.</param>
+        /// <param name="includePrerelease">If pre-release versions should be included.</param>
+        /// <returns> A dictionary where the key is the package version, and the value is the time it was published at.</returns>
+        public virtual Task<IDictionary<string, DateTime>> EnumerateVersionsWithPublishTimeAsync(PackageURL purl, bool useCache = true, bool includePrerelease = true)
```

Comment:
This kind of interface feels a bit narrow / not very extensible. Are there any cases where we would get a performance benefit from returning a version-to-datetime mapping, as opposed to a version-to-metadata-object mapping, where the metadata object can include datetime among other things?

I ask because it would not be a great user experience to have a bunch of separate version-to-something mapping methods, if/when we wanted to return more information like this.

Reply:
This could be my lack of knowledge, but I would imagine that there is a _slight_ efficiency increase if we are just trying to get a version-to-datetime mapping vs a version-to-metadata then selecting just the datetime from the metadata. But I could also see how there is such a negligible efficiency increase (if any at all) to the point where it doesn't really matter much anyways.

Reply:
I think any real improvement you’d get would be if the repository you’re talking to has an api specifically for this. If you have to fetch and decode the whole metadata and then just return the time stamp that’s not really saving anything over returning the whole parsed metadata - assuming the metadata with time is fetched in the same network call. 

That said, that now means that you need to implement that for every manager to get the benefit, which is even in the best case going to be slight I believe. 

The rub is, however, that we don’t have a standard metadata format (I believe?) so there isn’t a different place to slot in the DateTime. I do however think that it would be better to just return the metadata and let the caller look at the fields they want. a caller would then be able to just `.Select(x=> (x.version, x.datetime))` on the returned metadata if they wanted this behavior.

Reply:
There is a standard metadata format `PackageMetadata` which includes `UploadTime`, the bigger issue is that right now the metadata request only ever gets the values for a single version, we don't have a method for getting say a dictionary where the key is the version, and the value is the version specific metadata. Which would remove the need for this method and we would just replace it with something along the lines of the `.Select` you mentioned above.

Reply:
Ah I see. I think we can think of a way to structure that object so this data is in it.